xonsh/completer.py

```diff
@@ -257,3 +262,31 @@ def attr_complete(self, prefix, ctx):
             comp = prefix[:prelen-len(attr)] + rpl
             attrs.add(comp)
         return attrs
+
+    def _all_commands(self):
+        path = builtins.__xonsh_env__.get('PATH', None) or []
+        # did PATH change?
+        path_hash = hash(tuple(path))
+        cache_valid = path_hash == self._path_checksum
+        self._path_checksum = path_hash
+        # did aliases change?
+        al_hash = hash(tuple(builtins.aliases.keys()))
```

Comment:
I think that it would be more correct to have `hash(tuple(sorted(builtins.aliases.keys())))` here since key ordering is not guaranteed.  However, the worst that is going to happen is that it is going to trigger a cache miss.